core/trino-main/src/main/java/io/trino/operator/BigintGroupByHash.java

```diff
@@ -540,7 +583,101 @@ public GroupByIdBlock getResult()
             finished = true;
-            return new GroupByIdBlock(nextGroupId, blockBuilder.build());
+            return new GroupByIdBlock(nextGroupId, new LongArrayBlock(groupIds.length, Optional.empty(), groupIds));
+        }
+    }
+
+    private void batchedPutIfAbsent(int batchSize, int blockOffset, Block block, long[] groupIds, int groupIdOffset)
+    {
+        if (block.mayHaveNull()) {
+            batchedPutIfAbsentNullable(batchSize, blockOffset, block, groupIds, groupIdOffset);
+        }
+        else {
+            batchedPutIfAbsentNoNull(batchSize, blockOffset, block, groupIds, groupIdOffset);
+        }
+    }
+
+    private void batchedPutIfAbsentNullable(int batchSize, int blockOffset, Block block, long[] groupIds, int groupIdOffset)
+    {
+        int[] hashPositions = new int[batchSize];
+        for (int i = 0; i < batchSize; i++) {
+            if (!block.isNull(blockOffset + i)) {
+                hashPositions[i] = getHashPosition(BIGINT.getLong(block, blockOffset + i), mask);
+            }
+        }
+
+        for (int i = 0; i < batchSize; i++) {
+            if (!block.isNull(blockOffset + i)) {
+                groupIds[groupIdOffset + i] = this.groupIds[hashPositions[i]];
+            }
+            else {
+                groupIds[groupIdOffset + i] = NULL_GROUP_ID_MARKER;
+            }
+        }
+
+        for (int i = 0; i < batchSize; i++) {
+            if (groupIds[groupIdOffset + i] >= 0) {
+                long value = BIGINT.getLong(block, blockOffset + i);
+                long storedValue = values[hashPositions[i]];
+                boolean match = value == storedValue;
+                groupIds[groupIdOffset + i] = (groupIds[groupIdOffset + i] + 1) * (match ? 1 : 0) - 1;
+            }
+        }
+
+        if (nullGroupId < 0) { // Assigning new null group id adds another branch inside a loop so we want to do it only once
+            for (int i = 0; i < batchSize; i++) {
+                if (groupIds[groupIdOffset + i] == NULL_GROUP_ID_MARKER) {
+                    if (nullGroupId < 0) {
+                        // set null group id
+                        nullGroupId = nextGroupId++;
+                    }
+                    groupIds[groupIdOffset + i] = nullGroupId;
+                }
+                else if (groupIds[groupIdOffset + i] == -1) {
+                    groupIds[groupIdOffset + i] = putIfAbsent(BIGINT.getLong(block, blockOffset + i), hashPositions[i]);
+                }
+            }
+        }
+        else {
+            for (int i = 0; i < batchSize; i++) {
+                if (groupIds[groupIdOffset + i] == NULL_GROUP_ID_MARKER) {
+                    groupIds[groupIdOffset + i] = nullGroupId;
+                }
+                else if (groupIds[groupIdOffset + i] == -1) {
+                    groupIds[groupIdOffset + i] = putIfAbsent(BIGINT.getLong(block, blockOffset + i), hashPositions[i]);
+                }
+            }
+        }
+    }
```

Comment:
I'm thinking we could split it into smaller items that would reflect batch nature of the algorithm more, for example:
```
private void batchedPutIfAbsentNullable(Block block, int batchSize, int blockOffset, long[] groupIds, int groupIdOffset) {
  SelectedPositions nullPositions = SelectedPositions.withSize(MAX_BATCH_SIZE);
  SelectedPositions nonNullPositions = SelectedPositions.withSize(MAX_BATCH_SIZE);
  splitNullPositions(block, batchSize, blockOffset, nullPositions, nonNullPositions);
  getGroupIds(nonNullPositions, groupIds, groupIdOffset);
  SelectedPositions nonMatchingPositions = SelectedPositions.withSize(nonNullPositions.length());
  initialMatchPositions(block, blockOffset, groupIds, groupIdOffset, nonMatchingPositions);
  putRemainingPositions(block, blockOffset, groupIds, groupIdOffset, nonMatchingPositions);
  setNullGroupIds(nullPositions, groupIds, groupIdOffset);  
}
```

think about splitting the code into batch building blocks. I think we could apply similar pattern for other operators too. This is something I experimented with in https://github.com/trinodb/trino/pull/9232
(https://github.com/trinodb/trino/pull/9232/files#diff-f212a1fa9607f852e69facfa0c1cd6d853fd1584dbbe14896adf8f7d896f965aR273,
https://github.com/trinodb/trino/pull/9232/files#diff-f212a1fa9607f852e69facfa0c1cd6d853fd1584dbbe14896adf8f7d896f965aR165,
https://github.com/trinodb/trino/pull/9232/files#diff-f212a1fa9607f852e69facfa0c1cd6d853fd1584dbbe14896adf8f7d896f965aR331)
We could standarize how we internally handle batching within operators. I found that such approach with selected position is much more friendly entry point for join algorithm https://github.com/trinodb/trino/pull/9232/files#diff-ea07b50bba48518adc7b5f673faeb4dd1feaee08a4e2956f37817057a2f2ba31R123

Reply:
It looks like a nice API. However this does not fit the case at all. We do not need any list of selected positions to operate on. This will just introduce regressions.
I consider the idea OK, but it will not work here.

Reply:
> It looks like a nice API. However this does not fit the case at all. We do not need any list of selected positions to operate on.

Splitting null/non-null positions might be beneficial for performance (needs to be benchmarked). We do that in `io.trino.operator.output.PagePartitioner#partitionNullablePositions` for example

> This will just introduce regressions.

why? I suspect this might be better than running `if (isNull)` checks

Reply:
This PR does not touch the null handling, let's keep it simple

Reply:
what do you mean it doesn't touch nulls. There is `batchedPutIfAbsentNullable`

Reply:
It does not optimize nulls in any meaningfull way. `batchedPutIfAbsentNullable` works just a before. It is `batchedPutIfAbsentNoNull` that is optimized to run faster

Reply:
> It does not optimize nulls in any meaningfull way. batchedPutIfAbsentNullable works just a before

Not really, It's batched now, so why not to test if it actually helped or not?